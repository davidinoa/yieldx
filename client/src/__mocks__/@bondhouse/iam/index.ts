import {
  CreateInvestorMappingRequest,
  UserType,
  UserStatus
} from "@bondhouse/iam";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export function UsersApi() {}

UsersApi.prototype.getCurrentUser = jest.fn(() =>
  Promise.resolve({
    data: {
      user: {
        id: "auth0|5ec6119474dfa00cbf1d53cb",
        first: "Institution",
        last: "New User",
        userType: "ADMIN",
        email: "test4@bondhouse.com",
        terms: true,
        institutionId: "bf8a5597-9569-4510-babd-a6d9655e9f0d",
        createdAt: "2020-05-21T05:28:52.982380Z",
        updatedAt: "2020-05-21T05:28:52.982380Z",
        status: "ACTIVE",
        policies: []
      }
    }
  })
);

// eslint-disable-next-line @typescript-eslint/no-empty-function
export function PasswordResetTicketsApi() {
  return {};
}

PasswordResetTicketsApi.prototype.createPasswordResetTicket = jest.fn(() =>
  Promise.resolve({
    data: {
      success: true
    }
  })
);

// eslint-disable-next-line @typescript-eslint/no-empty-function
export function InstitutionInventoryMappingApi() {}

InstitutionInventoryMappingApi.prototype.getInstitutionInventoryMapping = jest.fn(
  () =>
    Promise.resolve({
      data: {
        inventoryMapping: {
          institutionId: "b17c644a-30e8-47fb-a26b-f6743786c3ba",
          inpaas: "inpaas",
          bestFit: "bestfit",
          assetExplorer: "assetexplorer"
        }
      }
    })
);
InstitutionInventoryMappingApi.prototype.createOrUpdateInstitutionInventoryMapping = jest.fn();

// eslint-disable-next-line @typescript-eslint/no-empty-function
export function InstitutionInvestorMappingApi() {}

InstitutionInvestorMappingApi.prototype.createInstitutionInvestorMapping = jest.fn(
  (
    institutionId: string,
    createInvestorMappingRequest?: CreateInvestorMappingRequest
  ) =>
    Promise.resolve({
      data: {
        investorMapping: {
          institutionId,
          investorId: createInvestorMappingRequest?.investorId || ""
        }
      }
    })
);

export { UserType, UserStatus };
