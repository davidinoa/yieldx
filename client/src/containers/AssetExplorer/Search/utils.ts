interface InGridFilter {
  filter: string | number;
  filterTo: number;
  filterType: string;
  type: string;
  dateFrom: string;
  dateTo: string;
  values: string[];
}

export interface AgGridFilters {
  [key: string]: any;
}

const getFilter = (filter: InGridFilter) => {
  /**
   *  The filter object from Ag Grid API. In the case of filtering `name` column.
   *  { filter: "Asset Name", filterType: "text", type: "contains"}
   */
  let searchFilter = {};
  const {
    filter: filterContent,
    filterType,
    type,
    filterTo,
    values,
    dateFrom,
    dateTo
  } = filter;
  if (filterType === "text") {
    switch (type) {
      case "contains":
        searchFilter = { containsText: filterContent };
        break;
      case "notContains":
        searchFilter = { doesNotContainText: filterContent };
        break;
      case "equals":
        searchFilter = { equalsText: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualText: filterContent };
        break;
      case "startsWith":
        searchFilter = { startsWithText: filterContent };
        break;
      case "endsWith":
        searchFilter = { endsWith: filterContent };
        break;
      default:
        searchFilter = { containsText: filterContent };
        break;
    }
  } else if (filterType === "number") {
    switch (type) {
      case "equals":
        searchFilter = { equalsNumber: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualNumber: filterContent };
        break;
      case "lessThan":
        searchFilter = { isLessThanNumber: filterContent };
        break;
      case "lessThanOrEqual":
        searchFilter = { isLessThanOrEqualToNumber: filterContent };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanNumber: filterContent };
        break;
      case "greaterThanOrEqual":
        searchFilter = { isGreaterThanOrEqualToNumber: filterContent };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToNumber: filterContent,
          isLessThanOrEqualToNumber: filterTo
        };
        break;
      default:
        searchFilter = { equalsNumber: filterContent };
        break;
    }
  } else if (filterType === "date") {
    const convertedDateFrom = new Date(dateFrom);
    const convertedDateTo = new Date(dateTo);
    switch (type) {
      case "equals":
        searchFilter = { equalsDate: convertedDateFrom };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanDate: convertedDateFrom };
        break;
      case "lessThan":
        searchFilter = { isLessThanDate: convertedDateFrom };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualDate: convertedDateFrom };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToDate: convertedDateFrom,
          isLessThanOrEqualToDate: convertedDateTo
        };
        break;
      default:
        searchFilter = { equalsDate: convertedDateFrom };
        break;
    }
  } else if (filterType === "set" && values) {
    // We may need to add `set` type Ag grid column filter params like cusip or isin
    // in the search request in the future.
    // searchFilter = { inText: values };
  }
  return searchFilter;
};

export const convertInGridFilters = (filters: AgGridFilters) => {
  const convertedFilters: { [index: string]: any } = {};
  Object.keys(filters).forEach(key => {
    const inGridFilter = filters[key];
    if (key === "minimumQuantity") key = "bond.minimumDenomination";
    convertedFilters[key] = getFilter(inGridFilter);
  });

  return convertedFilters;
};
