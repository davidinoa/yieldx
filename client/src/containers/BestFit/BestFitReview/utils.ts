import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
