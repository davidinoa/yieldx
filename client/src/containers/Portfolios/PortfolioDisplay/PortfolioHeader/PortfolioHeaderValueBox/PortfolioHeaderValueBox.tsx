import React, { Dispatch, SetStateAction, useState } from "react";
import { useHistory } from "react-router-dom";
import { PortfolioType } from "@bondhouse/income-portfolios";
import YieldXPortfolio from "models/YieldXPortfolio";
import { PortfolioStatus } from "@bondhouse/position-management";
import toInputPortfolio from "util/toInputPortfolio";
import LiveHeaderOptions from "./LiveHeaderOptions/LiveHeaderOptions";
import { Value, ValueContainer } from "../PortfolioHeader.styles";
import ExecutionHoursModal from "../../Modals/ExecutionHoursModal/ExecutionHoursModal";
import PortfolioValue from "./PortfolioValue/PortfolioValue";
import ExternalHeaderOptions from "./ExternalHeaderOptions/ExternalHeaderOptions";
import LiveProfitLossValues from "./LiveProfitLossValues/LiveProfitLossValues";

interface Props {
  portfolio: YieldXPortfolio;
  isReview?: boolean;
  investorId?: string;
  setShowInvestModal: Dispatch<SetStateAction<boolean>>;
}

export default function PortfolioHeaderValueBox({
  portfolio,
  investorId,
  isReview,
  setShowInvestModal
}: Props) {
  const history = useHistory();
  const [showHoursModal, setShowHoursModal] = useState(false);
  const isLive = portfolio.getType() === PortfolioType.LIVE;
  // TODO: Remove this redundant uppercase check once enum casing issue is resolved
  const isTerminated =
    portfolio.getStatus() === PortfolioStatus.TERMINATED ||
    portfolio.getStatus().toUpperCase() === "TERMINATED";
  const isPaper = portfolio.getType() === PortfolioType.EXTERNAL;
  return (
    <>
      <Value>
        <ValueContainer>
          <PortfolioValue value={portfolio.totalMarketValue} />
          {!isReview && isLive && (
            <LiveHeaderOptions investorId={investorId} portfolio={portfolio} />
          )}
          {isLive && !isTerminated && (
            <LiveProfitLossValues
              portfolioInput={toInputPortfolio(portfolio)}
            />
          )}
          {isPaper && !isReview && (
            <ExternalHeaderOptions
              portfolio={portfolio}
              setShowInvestModal={setShowInvestModal}
              setShowHoursModal={setShowHoursModal}
            />
          )}
        </ValueContainer>
      </Value>
      <ExecutionHoursModal
        isOpen={showHoursModal}
        onHide={() => setShowHoursModal(false)}
        onSave={() => {
          setShowHoursModal(false);
          history.push("#save");
        }}
      />
    </>
  );
}
