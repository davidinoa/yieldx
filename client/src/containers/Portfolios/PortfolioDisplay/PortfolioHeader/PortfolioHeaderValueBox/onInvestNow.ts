import * as H from "history";
import dayjs from "dayjs";
import { Dispatch, SetStateAction } from "react";
import { SetterOrUpdater } from "recoil";
import { AppType } from "@bondhouse/position-management";
import { isTradingOpen } from "containers/Portfolios/PortfolioDisplay/Modals/ExecutionHoursModal/ExecutionHoursModal";
import YieldXPortfolio from "models/YieldXPortfolio";
import { PortfolioUIState } from "services/PortfolioService";
import { calculateDaysSince } from "util/calculateDaysSince";

export default function onInvestNow({
  history,
  portfolio,
  setShowHoursModal,
  setShowInvestModal,
  setPortfolioState
}: {
  history: H.History<unknown>;
  portfolio: YieldXPortfolio;
  setShowInvestModal: Dispatch<SetStateAction<boolean>>;
  setShowHoursModal: Dispatch<SetStateAction<boolean>>;
  setPortfolioState: SetterOrUpdater<PortfolioUIState>;
}) {
  const tradingOpen = isTradingOpen(dayjs());
  const isProposalStale =
    calculateDaysSince(portfolio.getUpdatedAt() ?? portfolio.getCreatedAt()) >=
    1;
  if (
    process.env.NODE_ENV === "development" ||
    window.location.host.split(".")[0] === "dev"
  ) {
    if (portfolio.getAppType() === AppType.INPAAS && !isProposalStale) {
      setPortfolioState({
        status: "executing",
        currentPortfolio: portfolio
      });
      history.push("/execute-portfolio");
    } else setShowInvestModal(true);
  } else if (process.env.NODE_ENV === "production" && tradingOpen)
    setShowInvestModal(true);
  else if (process.env.NODE_ENV === "production" && !tradingOpen)
    setShowHoursModal(true);
}
