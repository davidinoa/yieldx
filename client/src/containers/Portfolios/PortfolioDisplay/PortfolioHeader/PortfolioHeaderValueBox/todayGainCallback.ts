import { PortfolioAnalyzerApi } from "@bondhouse/rover-portfolio-analyzer";
import YieldXPortfolio from "models/YieldXPortfolio";

const analyzerApi = new PortfolioAnalyzerApi();

export default function todayGainCallback(portfolio: YieldXPortfolio) {
  return async () => {
    const {
      data: { marketValueReturn }
    } = await analyzerApi.calculateIntradayReturn({
      portfolio: portfolio.asPmsPortfolio()
    });
    return marketValueReturn;
  };
}
