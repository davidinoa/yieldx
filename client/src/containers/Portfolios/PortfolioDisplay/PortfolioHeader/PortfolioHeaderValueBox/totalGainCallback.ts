import { PortfolioAnalyzerApi } from "@bondhouse/rover-portfolio-analyzer";
import dayjs from "dayjs";
import YieldXPortfolio from "../../../../../models/YieldXPortfolio";

const analyzerApi = new PortfolioAnalyzerApi();

export default function totalGainCallback(portfolio: YieldXPortfolio) {
  return async () => {
    const {
      data: {
        estimatedPerformance: { timeSeries }
      }
    } = await analyzerApi.calculateReturns({
      startDate: dayjs(portfolio.getCreatedAt()).format("YYYY-MM-DD"),
      endDate: dayjs().format("YYYY-MM-DD"),
      portfolio: portfolio.asPmsPortfolio()
    });
    return timeSeries[timeSeries.length - 1]?.cumulativeReturn ?? 0;
  };
}
