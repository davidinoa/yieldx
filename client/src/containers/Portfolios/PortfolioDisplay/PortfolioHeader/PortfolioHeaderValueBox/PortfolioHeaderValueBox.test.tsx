import React from "react";
import { render } from "test/utils";
import { mockMultiAssetPortfolio } from "__mocks__/models/mockYieldXPortfolio";
import { PortfolioType } from "@bondhouse/income-portfolios";
import { userState } from "services/userState";
import { UserType } from "@bondhouse/iam";
import PortfolioHeaderValueBox from "./PortfolioHeaderValueBox";

describe("PortfolioHeaderValueBox", () => {
  test("Should render", () => {
    const { container } = render(
      <PortfolioHeaderValueBox
        portfolio={mockMultiAssetPortfolio}
        setShowInvestModal={jest.fn()}
      />
    );
    expect(container).toBeInTheDocument();
  });

  test("Should render live portfolio options when portfolioType is LIVE", () => {
    const { asFragment } = render(
      <PortfolioHeaderValueBox
        portfolio={mockMultiAssetPortfolio}
        setShowInvestModal={jest.fn()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("Should render external portfolio options when portfolioType is EXTERNAL", () => {
    const clone = mockMultiAssetPortfolio.clone();
    clone.setType(PortfolioType.EXTERNAL)
    const { asFragment } = render(
      <PortfolioHeaderValueBox
        portfolio={clone}
        setShowInvestModal={jest.fn()}
      />, {
        initializeState: ({set}: {set: any}) => set(userState, {
          institutionId: "mock-institution-id",
          id: "mock-user-id",
          userType: UserType.ADMIN,
          first: "mock",
          last: "user",
          policies: [],
          email: "mock@email.com",
          terms: true
        })
      }
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
