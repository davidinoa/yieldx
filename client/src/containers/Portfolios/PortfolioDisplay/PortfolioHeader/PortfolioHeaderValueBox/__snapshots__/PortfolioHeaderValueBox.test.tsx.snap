// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortfolioHeaderValueBox Should render external portfolio options when portfolioType is EXTERNAL 1`] = `
<DocumentFragment>
  .c0 {
  border: 1px solid var(--border-new);
  border-radius: 0.35rem;
  padding: 0.5rem 0.75rem;
}

.c1 {
  background: rgba(0,0,255,0.3);
  border-radius: 0.35rem 0.35rem 0 0;
  background: var(--dark);
  padding: 0.35rem;
}

.c1 h3 {
  color: var(--lightest);
  text-align: center;
  line-height: 1em;
  margin: 0;
  font-size: 0.75rem;
}

.c2 {
  position: relative;
  padding-top: 0.85rem;
}

.c2 span {
  display: block;
  text-align: center;
  font-size: 1.7rem;
  font-weight: bold;
  padding-bottom: 0.45rem;
}

.c2 span > sup {
  font-size: 0.55em;
  line-height: 22px;
  padding: 0.25rem;
}

.c3 {
  border-top: solid 1px var(--border-new);
  text-align: center;
}

.c3 button,
.c3 a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: none;
  background: none;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  margin: 0.5rem 0 0;
  color: var(--lightest);
  width: auto;
  position: relative;
  padding: 0 0.65rem;
  font-size: 0.7rem;
}

.c3 button.invest-cta,
.c3 a.invest-cta {
  background: var(--primary);
  color: white;
  border-radius: 0.15rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 0.15rem 0.75rem;
  font-weight: bold;
}

.c3 button:last-of-type,
.c3 a:last-of-type {
  border: none;
}

.c3 span {
  margin: 0.5rem 0 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-right: solid 1px var(--border-new);
}

@media (min-width:576px) {

}

@media (min-width:1600px) {

}

@media (min-width:1600px) {
  .c1 h3 {
    font-size: 0.75rem;
  }
}

@media (min-width:1600px) {
  .c2 {
    padding-top: 1.25rem;
  }
}

@media (min-width:1600px) {
  .c2 span {
    font-size: 2rem;
  }
}

@media (min-width:1600px) {
  .c3 button,
  .c3 a {
    padding: 0 1rem;
    font-size: 0.8rem;
  }
}

<div
    class="col-8 col-xxl-7"
  >
    <div
      class="c0 container-fluid"
    >
      <div
        class="row"
      >
        <div
          class="c1 col-24"
        >
          <h3>
            PORTFOLIO VALUE
          </h3>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="c2 col"
        >
          <span>
            $61,722
            <sup>
              .00
            </sup>
          </span>
        </div>
      </div>
      <div
        class="row"
        data-testid="External-Portfolio-Header-Options"
      >
        <div
          class="c3 col-24 d-flex flex-row justify-content-around"
        >
          <button
            data-testid="rebalance-edit-button"
            type="button"
          >
            Edit positions
          </button>
          <span />
          <button
            class=""
            data-testid="delete-button"
            type="button"
          >
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PortfolioHeaderValueBox Should render live portfolio options when portfolioType is LIVE 1`] = `
<DocumentFragment>
  .c7 {
  margin: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  font-size: 0.5rem;
  color: var(--lightest);
}

.c6 {
  height: 100%;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  width: 100%;
  height: 100%;
}

.c0 {
  border: 1px solid var(--border-new);
  border-radius: 0.35rem;
  padding: 0.5rem 0.75rem;
}

.c1 {
  background: rgba(0,0,255,0.3);
  border-radius: 0.35rem 0.35rem 0 0;
  background: var(--dark);
  padding: 0.35rem;
}

.c1 h3 {
  color: var(--lightest);
  text-align: center;
  line-height: 1em;
  margin: 0;
  font-size: 0.75rem;
}

.c2 {
  position: relative;
  padding-top: 0.85rem;
}

.c2 span {
  display: block;
  text-align: center;
  font-size: 1.7rem;
  font-weight: bold;
  padding-bottom: 0.45rem;
}

.c2 span > sup {
  font-size: 0.55em;
  line-height: 22px;
  padding: 0.25rem;
}

.c4 {
  padding-top: 0.85rem;
  border-top: 1px solid var(--border-new);
}

.c4 span.label {
  display: inherit;
  font-weight: bold;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.2em;
  -moz-letter-spacing: 0.2em;
  -ms-letter-spacing: 0.2em;
  letter-spacing: 0.2em;
  font-weight: bold;
  font-family: var(--font-primary);
  color: var(--subtext-3);
  font-size: 0.6rem;
}

.c4 span.value {
  display: inherit;
  text-align: center;
  font-family: var(--font-secondary);
  font-weight: normal;
  font-size: 0.7rem;
}

.c4 span.value.positive {
  color: var(--active);
}

.c4 span.value.negative {
  color: var(--issue);
}

.c4 .divider {
  border-right: 1px solid var(--border-new);
}

.c3 {
  right: 5%;
  top: 25%;
  -webkit-transform: translate(0%,75%);
  -ms-transform: translate(0%,75%);
  transform: translate(0%,75%);
  border: none;
  background: none;
  position: absolute;
  width: 20px;
  height: 20px;
  background-image: radial-gradient( circle, var(--light) 2px, transparent 2.5px );
  background-size: 100% 33.33%;
}

@media (min-width:576px) {

}

@media (min-width:1600px) {

}

@media (min-width:1600px) {
  .c1 h3 {
    font-size: 0.75rem;
  }
}

@media (min-width:1600px) {
  .c2 {
    padding-top: 1.25rem;
  }
}

@media (min-width:1600px) {
  .c2 span {
    font-size: 2rem;
  }
}

@media (min-width:1600px) {
  .c4 {
    padding-top: 1.25rem;
  }
}

@media (min-width:1600px) {
  .c4 span.label {
    font-size: 0.75rem;
  }
}

@media (min-width:1600px) {
  .c4 span.value {
    font-size: 0.85rem;
  }
}

<div
    class="col-8 col-xxl-7"
  >
    <div
      class="c0 container-fluid"
    >
      <div
        class="row"
      >
        <div
          class="c1 col-24"
        >
          <h3>
            PORTFOLIO VALUE
          </h3>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="c2 col"
        >
          <span>
            $61,722
            <sup>
              .00
            </sup>
          </span>
        </div>
      </div>
      <button
        class="c3"
        data-testid="more-button"
        type="button"
      />
      <div
        class="row"
      >
        <div
          class="c4 col-24 d-flex justify-content-around"
        >
          <div
            class="d-inline-block performance"
          >
            <span
              class="label"
            >
              today gain/loss:
            </span>
            <span>
              <div
                class="c5"
                height="100%"
              >
                <div
                  class="c6"
                >
                  <div
                    class="c7 spinner-border"
                    data-testid="spinner"
                  />
                </div>
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
